### PR TITLE
feat(component): TreeView

### PR DESCRIPTION
**Pull request by michaelkro**
**What**: Limited Function TreeView

**Link to Storybook**: https://michaelkro.github.io/patternfly-react/?selectedKind=TreeView&selectedStory=TreeView&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybooks%2Fstorybook-addon-knobs

**Comments**:  Introductory TreeView pattern based on http://www.patternfly.org/pattern-library/widgets/#bootstrap-tree-view.  Initial functionality includes

* Rendering the tree
* Expanding and collapsing nodes with children
* Clicking on a node with its `selectable` flag set to true fires a callback (provided to component as a prop) that has access to the selected node as its first argument
* (Configurable) Display the tree expanded by default
* Highlighting on hover and select

Plans for next iteration
* Lift state from individual nodes

### CODE REVIEW

File: src/components/TreeView/TreeViewNode.js [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    super(props);
+
+    // A node can be set to be expanded by default
+    this.state = {
+      expanded:
+        (props.node.hasOwnProperty('state') && props.node.state.expanded) ||
+        false
+    };
+
+    this.toggleExpand = this.toggleExpand.bind(this);
```
Comment (priley86) [inline src/components/TreeView/TreeViewNode.js]: we have bind methods for these now...

import { bindMethods } from '../../common/helpers';

File: src/components/TreeView/TreeViewNode.js [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+  // Collapse the current node if any of its parents is collapsed. This should
+  // only fire for nodes that are level 2 or greater
+  componentWillReceiveProps(nextProps) {
+    if (!nextProps.visible && nextProps.level > 1) {
+      this.setState(() => ({ expanded: false }));
+    }
+  }
+
+  toggleExpand() {
+    this.setState(prevState => ({ expanded: !prevState.expanded }));
```
Comment (priley86) [inline src/components/TreeView/TreeViewNode.js]: for the POC... i think using component state is fine... but preference long term is to lift state to redux or use the `controlled()` pattern introduced.

**Comment by priley86** [comment]
nice job turning this around so quickly! I think we can test this on the integration side soon...  maybe you can introduce a `MockTreeViewContainer` or `MockTreeViewManager` to demonstrate that next?

File: src/components/TreeView/TreeViewNode.js (hunk 2 above)
Comment (michaelkro) [reply src/components/TreeView/TreeViewNode.js]: yes, I was a little confused when I was building this out, but now I think I understand what I need to do.  I'm starting to explore a stateless version now to prepare for the conversion

**Comment by michaelkro** [comment]
Thanks @priley86 :)

Yes, I'll put something together based on the POC mocks and run it by everyone to see if we have what we need

**Comment by priley86** [comment]
this looks great to me!

@michaelkro @serenamarie125 one question here: how do we plan to remove individual tree nodes? It looks like the `MockTreeViewManager` example allows you to remove individual parent nodes (good), but it might be nice to show when they are selected? I know we have some patterns w/ Treeview Checkboxes/Treeview highlight select in PF Core, do we want to tackle those yet?

**Comment by michaelkro** [comment]
@priley86 I can try using the styles defined here:

https://github.com/patternfly/patternfly/blob/master/src/sass/converted/patternfly/_bootstrap-treeview.scss#L74-L88

But would that be a bit abusive, since the class names seem to indicate that there is a dropdown?

File: src/components/TreeView/TreeViewNode.js [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+      <React.Fragment>
+        <li className={classes} onClick={this.handleSelect}>
+          <TreeViewIndents level={level} />
+          <TreeViewExpand
+            nodes={node.nodes}
+            expanded={expanded}
+            toggleExpand={this.toggleExpand}
+          />
+          <TreeViewIcon icon={node.icon} />
+          {node.text}
```
Comment (jeff-phillips-18) [inline src/components/TreeView/TreeViewNode.js]: Would be nice to toggleExpand on a double click here.

File: src/components/TreeView/TreeView.js [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+      </ul>
+    </div>
+  );
+};
+
+TreeView.propTypes = {
+  nodes: PropTypes.array,
+  selectNode: PropTypes.func,
+  highlightOnHover: PropTypes.bool,
+  highlightOnSelect: PropTypes.bool
```
Comment (jeff-phillips-18) [inline src/components/TreeView/TreeView.js]: Comments for the storybook would be helpful here. What fields should the nodes have?

File: src/components/TreeView/stories/treeViewManagerStory.js [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+  </div>
+);
+
+const treeViewManagerStory = stories => {
+  stories.addWithInfo('Mock TreeView Manager', '', () => {
+    const story = <MockTreeViewManager />;
+
+    return template('Mock TreeView Manager', story, description);
+  });
+};
```
Comment (jeff-phillips-18) [inline src/components/TreeView/stories/treeViewManagerStory.js]: The storybook would be much more helpful if it included the code necessary to create this mock. Currently, there isn't much there that is of any help.

File: src/components/TreeView/TreeView.js (hunk 4 above)
Comment (jeff-phillips-18) [reply src/components/TreeView/TreeView.js]: There should be defaults for highlightOnHover and highlightOnSelect

File: src/components/TreeView/stories/basicTreeView.js [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+import { basicData } from '../__mocks__/data';
+
+const basicTreeView = stories => {
+  stories.addWithInfo('Basic TreeView', '', () => {
+    const story = <TreeView nodes={basicData} />;
+
+    return template('Basic TreeView', story);
+  });
+};
+
```
Comment (jeff-phillips-18) [inline src/components/TreeView/stories/basicTreeView.js]: Would be good to have knobs for the highlight options.

File: src/components/TreeView/TreeView.js (hunk 4 above)
Comment (michaelkro) [reply src/components/TreeView/TreeView.js]: @jeff-phillips-18 how do you feel about me adding a story like this

fullscreen_2_7_18__12_20_pm

with regard to https://github.com/patternfly/patternfly-react/pull/196#discussion_r166654514?

File: src/components/TreeView/TreeViewNode.js (hunk 3 above)
Comment (michaelkro) [reply src/components/TreeView/TreeViewNode.js]: @jeff-phillips-18 I'm happy to implement that, but that strays from the pf-core example.  Should I maybe try to wire it up as an option?

File: src/components/TreeView/TreeView.js (hunk 4 above)
Comment (jeff-phillips-18) [reply src/components/TreeView/TreeView.js]: Is there a way to do it via comments had have it look decent in the props table?

Comment (michaelkro) [reply src/components/TreeView/TreeView.js]: oohhh, ok, sorry @jeff-phillips-18, I'm picking up what you're putting down now.  Yes, I'll look into that and including the source for the other storybook example.  Thanks for pointing that out!

**Comment by priley86** [comment]
@michaelkro one other minor thing I am noticing now... our docs on the PF org website linked via the storybook say "Bootstrap Tree View" and link to "patternfly-bootstrap-treeview". 

@LHinson @serenamarie125 maybe we can change the language on PF org website soon? The same would apply for the "Bootstrap Switch" it looks like...

File: src/components/TreeView/stories/treeViewManagerStory.js (hunk 5 above)
Comment (michaelkro) [reply src/components/TreeView/stories/treeViewManagerStory.js]: @jeff-phillips-18 @priley86 I created this mock mainly as an integration test for the v2v wizard. Is it appropriate to include this in the storybook, or should I remove it?

I've fixed up the main example with knobs, source, and prop tables.

I'm happy to add more as well, just not sure if we want to keep this particular one

**Comment by priley86** [comment]
@michaelkro for the initial implementation, very nicely done 👍

**Pull request merged**